# Data
Before getting started, let's take a look at the provided data

## Provided datasets

### GILGFVFTL_data
This dataset contains positive examples of TCR Cells (alpha and beta parts (sometimes both, sometimes only one of them)) that react with th GILGFVFTL epitope.

In [23]:
# Load data/GILGFVFTL_data.tsv in a dataframe
import pandas as pd
df_g = pd.read_csv('data/GILGFVFTL_data.tsv') #, sep='\t') # Hmm, for some reason it has a .tsv extension but is actually a .csv file
df_g.head()

Unnamed: 0 GeneA        CDR3_alfa       TRAV       TRAJ      MHC A_alfa  \
0           0   TRA     CAGAGSQGNLIF  TRAV27*01  TRAJ42*01  HLA-A*02:01:48   
1           1   TRA     CAGAGSQGNLIF  TRAV27*01  TRAJ42*01  HLA-A*02:01:48   
2           2   TRA     CAGAGSQGNLIF  TRAV27*01  TRAJ42*01  HLA-A*02:01:48   
3           3   TRA     CAGAGSQGNLIF  TRAV27*01  TRAJ42*01  HLA-A*02:01:48   
4           4   TRA  CAGPGGSSNTGKLIF  TRAV35*01  TRAJ37*01  HLA-A*02:01:48   

     Epitope  Score_alfa GeneB      CDR3_beta       TRBV        TRBJ  \
0  GILGFVFTL         3.0   TRB  CASSSRSSYEQYF  TRBV19*01  TRBJ2-7*01   
1  GILGFVFTL         3.0   TRB  CASSSRSSYEQYF  TRBV19*01  TRBJ2-7*01   
2  GILGFVFTL         3.0   TRB  CASSSRSSYEQYF  TRBV19*01  TRBJ2-7*01   
3  GILGFVFTL         3.0   TRB  CASSSRASYEQYF  TRBV19*01  TRBJ2-7*01   
4  GILGFVFTL         3.0   TRB  CASSLIYPGELFF  TRBV27*01  TRBJ2-2*01   

       MHC A_beta  Epitope.1  Score_beta  
0  HLA-A*02:01:48  GILGFVFTL         3.0  
1  HLA-A*02:01:48  GILGFVFTL         3.0  
2  HLA-A*02:01:48  GILGFVFTL         3.0  
3  HLA-A*02:01:48  GILGFVFTL         3.0  
4  HLA-A*02:01:48  GILGFVFTL         3.0

In [24]:
# the allele is also given here (e.g. *01), which is not in the other datasets, so let's remove the * and everying after it in TRAV, TRAJ, TRBV, TRBJ
for col in ['TRAJ', 'TRAV', 'TRBV', 'TRBJ']:
    df_g[col] = df_g[col].str.split('*').str[0]
df_g.head()

Unnamed: 0 GeneA        CDR3_alfa    TRAV    TRAJ      MHC A_alfa  \
0           0   TRA     CAGAGSQGNLIF  TRAV27  TRAJ42  HLA-A*02:01:48   
1           1   TRA     CAGAGSQGNLIF  TRAV27  TRAJ42  HLA-A*02:01:48   
2           2   TRA     CAGAGSQGNLIF  TRAV27  TRAJ42  HLA-A*02:01:48   
3           3   TRA     CAGAGSQGNLIF  TRAV27  TRAJ42  HLA-A*02:01:48   
4           4   TRA  CAGPGGSSNTGKLIF  TRAV35  TRAJ37  HLA-A*02:01:48   

     Epitope  Score_alfa GeneB      CDR3_beta    TRBV     TRBJ  \
0  GILGFVFTL         3.0   TRB  CASSSRSSYEQYF  TRBV19  TRBJ2-7   
1  GILGFVFTL         3.0   TRB  CASSSRSSYEQYF  TRBV19  TRBJ2-7   
2  GILGFVFTL         3.0   TRB  CASSSRSSYEQYF  TRBV19  TRBJ2-7   
3  GILGFVFTL         3.0   TRB  CASSSRASYEQYF  TRBV19  TRBJ2-7   
4  GILGFVFTL         3.0   TRB  CASSLIYPGELFF  TRBV27  TRBJ2-2   

       MHC A_beta  Epitope.1  Score_beta  
0  HLA-A*02:01:48  GILGFVFTL         3.0  
1  HLA-A*02:01:48  GILGFVFTL         3.0  
2  HLA-A*02:01:48  GILGFVFTL         3.0  
3  HLA-A*02:01:48  GILGFVFTL         3.0  
4  HLA-A*02:01:48  GILGFVFTL         3.0

In [25]:
# Get per column the number of different values
df_g.nunique()

Unnamed: 0    5192
GeneA            1
CDR3_alfa     1093
TRAV            42
TRAJ            48
MHC A_alfa       3
Epitope          1
Score_alfa       3
GeneB            1
CDR3_beta     3473
TRBV            47
TRBJ            13
MHC A_beta       3
Epitope.1        1
Score_beta       4
dtype: int64

In [26]:
# Number of genes containing alpha (column GeneA = TRA)
alpha_count_g = df_g[df_g['GeneA'] == 'TRA'].shape[0]
f"Number of genes containing alpha: {alpha_count_g} ({alpha_count_g/df_g.shape[0]*100:.2f}%)"

'Number of genes containing alpha: 2161 (41.62%)'

In [27]:
beta_count_g = df_g[df_g['GeneB'] == 'TRB'].shape[0]
f"Number of genes containing beta: {beta_count_g} ({beta_count_g/df_g.shape[0]*100:.2f}%)"

'Number of genes containing beta: 5190 (99.96%)'

In [28]:
# Values to ignore (for now?) are MHC A_alfa (always the same), Score_alfa (confidence score of alfa), Score_beta (confidence score of beta) -> Drop those columns
df_g = df_g.drop(columns=['MHC A_alfa', 'Score_alfa', 'MHC A_beta', 'Score_beta'])
df_g.head()

Unnamed: 0 GeneA        CDR3_alfa    TRAV    TRAJ    Epitope GeneB  \
0           0   TRA     CAGAGSQGNLIF  TRAV27  TRAJ42  GILGFVFTL   TRB   
1           1   TRA     CAGAGSQGNLIF  TRAV27  TRAJ42  GILGFVFTL   TRB   
2           2   TRA     CAGAGSQGNLIF  TRAV27  TRAJ42  GILGFVFTL   TRB   
3           3   TRA     CAGAGSQGNLIF  TRAV27  TRAJ42  GILGFVFTL   TRB   
4           4   TRA  CAGPGGSSNTGKLIF  TRAV35  TRAJ37  GILGFVFTL   TRB   

       CDR3_beta    TRBV     TRBJ  Epitope.1  
0  CASSSRSSYEQYF  TRBV19  TRBJ2-7  GILGFVFTL  
1  CASSSRSSYEQYF  TRBV19  TRBJ2-7  GILGFVFTL  
2  CASSSRSSYEQYF  TRBV19  TRBJ2-7  GILGFVFTL  
3  CASSSRASYEQYF  TRBV19  TRBJ2-7  GILGFVFTL  
4  CASSLIYPGELFF  TRBV27  TRBJ2-2  GILGFVFTL

### background
This dataset contains some alpha and beta parts from TCR cells, without any epitope. You can use them as negative dataset (by filtering out the alphas and betas that occur in the positive dataset).

In [29]:
import pandas as pd
df_b = pd.read_csv('data/background.tsv', sep='\t') # this is actually a tsv
df_b.head()

CDR3_alfa         TRAV    TRAJ         CDR3_beta      TRBV     TRBJ
0   CAYGSTYNTDKLIF  TRAV38-2DV8  TRAJ34  CASSQEGSGVTDTQYF   TRBV4-3  TRBJ2-3
1  CILRDEGGGADGLTF     TRAV26-2  TRAJ45   CSAREGLAEFNEQFF  TRBV20-1  TRBJ2-1
2     CAGGSGYSTLTF     TRAV12-2  TRAJ11      CASSLGHYGYTF  TRBV12-3  TRBJ1-2
3  CAVRDLIVGANNLFF        TRAV3  TRAJ36    CASSQSFRDDEQYF    TRBV18  TRBJ2-7
4    CATYGGSQGNLIF       TRAV21  TRAJ42    CASSQAVGYNEQFF   TRBV4-1  TRBJ2-1

In [30]:
df_b.nunique()

CDR3_alfa    429396
TRAV             56
TRAJ             56
CDR3_beta    474559
TRBV             61
TRBJ             13
dtype: int64

## Combined Dataset

In [31]:
positive_dataset = df_g
negative_dataset = df_b

In [32]:
# Drop the rows from the negative dataset where CDR3_alpha or CDR3_beta occur in the positive dataset
# Note: Check seperately or together? (e.g. alpha in positive, but alpha + beta combination not)
negative_dataset = negative_dataset[~negative_dataset['CDR3_alfa'].isin(positive_dataset['CDR3_alfa'])]
negative_dataset = negative_dataset[~negative_dataset['CDR3_beta'].isin(positive_dataset['CDR3_beta'])]

In [33]:
# get the percentages of df_g containing alpha, beta and both
alpha_only_count_g = df_g[(df_g['GeneA'] == 'TRA') & (df_g['GeneB'] != 'TRB')].shape[0]
beta_only_count_g = df_g[(df_g['GeneA'] != 'TRA') & (df_g['GeneB'] == 'TRB')].shape[0]
both_count_g = df_g[(df_g['GeneA'] == 'TRA') & (df_g['GeneB'] == 'TRB')].shape[0]
non_count_g = df_g[(df_g['GeneA'] != 'TRA') & (df_g['GeneB'] != 'TRB')].shape[0]
print(f"Number of genes containing alpha only: {alpha_only_count_g} ({alpha_only_count_g/df_g.shape[0]*100:.2f}%)")
print(f"Number of genes containing beta only: {beta_only_count_g} ({beta_only_count_g/df_g.shape[0]*100:.2f}%)")
print(f"Number of genes containing both: {both_count_g} ({both_count_g/df_g.shape[0]*100:.2f}%)")
print(f"Number of genes containing neither: {non_count_g} ({non_count_g/df_g.shape[0]*100:.2f}%)")

Number of genes containing alpha only: 2 (0.04%)
Number of genes containing beta only: 3031 (58.38%)
Number of genes containing both: 2159 (41.58%)
Number of genes containing neither: 0 (0.00%)


In [34]:
def sample_and_drop(df, n):
    """Sample n rows from df and drop them from df"""
    # src: https://stackoverflow.com/questions/39835021/pandas-random-sample-with-remove
    df_subset = df.sample(n)
    df.drop(df_subset.index, inplace=True)
    return df_subset

In [35]:
# Test the sample and drop function
df_ex = pd.DataFrame({'a': [1,2,3,4,5,6,7,8,9,10], 'b': [1,2,3,4,5,6,7,8,9,10]})
# get the number of rows in df_ex
rows_before = df_ex.shape[0]
n = 2
df_subset = sample_and_drop(df_ex, n)
rows_after = df_ex.shape[0]
assert rows_before - rows_after == n
df_subset

a  b
5  6  6
1  2  2

In [36]:
import numpy as np

# now sample the same amount of negative examples from the negative dataset
alpha_only_rows_b = sample_and_drop(negative_dataset, alpha_only_count_g)
# Set CDR3_beta, TRBV and TRBJ to NaN
alpha_only_rows_b['CDR3_beta'] = np.nan
alpha_only_rows_b['TRBV'] = np.nan
alpha_only_rows_b['TRBJ'] = np.nan

# Is this the
beta_only_rows_b = sample_and_drop(negative_dataset, beta_only_count_g)
# Set CDR3_alpha, TRAV and TRAJ to NaN
beta_only_rows_b['CDR3_alfa'] = np.nan
beta_only_rows_b['TRAV'] = np.nan
beta_only_rows_b['TRAJ'] = np.nan

both_rows_b = sample_and_drop(negative_dataset, both_count_g)

non_rows_b = sample_and_drop(negative_dataset, non_count_g)
# Set CDR3_alpha, TRAV,  TRAJ, CDR3_beta, TRBV, TRBJ to NaN
non_rows_b['CDR3_alfa'] = np.nan
non_rows_b['TRAV'] = np.nan
non_rows_b['TRAJ'] = np.nan
non_rows_b['CDR3_beta'] = np.nan
non_rows_b['TRBV'] = np.nan
non_rows_b['TRBJ'] = np.nan

In [37]:
# combine samples in new dataframe
negative_dataset_same_proportion = pd.concat([alpha_only_rows_b, beta_only_rows_b, both_rows_b, non_rows_b])
# Check whether number of rows is same as in positive dataset
assert negative_dataset_same_proportion.shape[0] == positive_dataset.shape[0]

In [38]:
# count the number of duplicates in the positive dataset
positive_dataset_duplicates = positive_dataset[positive_dataset.duplicated()]
print(f"Number of duplicates in positive dataset: {positive_dataset_duplicates.shape[0]}")

# count the number of duplicates in the negative dataset
negative_dataset_duplicates = negative_dataset_same_proportion[negative_dataset_same_proportion.duplicated()]
print(f"Number of duplicates in negative dataset: {negative_dataset_duplicates.shape[0]}")

# Count the number of duplicates in the combined dataset
combined_dataset_duplicates = pd.concat([positive_dataset, negative_dataset_same_proportion])
combined_dataset_duplicates = combined_dataset_duplicates[combined_dataset_duplicates.duplicated()]
print(f"Number of duplicates in combined dataset: {combined_dataset_duplicates.shape[0]}")

Number of duplicates in positive dataset: 0
Number of duplicates in negative dataset: 0
Number of duplicates in combined dataset: 0


In [39]:
# Add column 'reaction' 0 for negative_dataset_same_proportion, 1 for positive_dataset
positive_dataset['reaction'] = 1
negative_dataset_same_proportion['reaction'] = 0

In [40]:
# Keep only the columns of the positive dataset that are in the negative datasets
positive_dataset = positive_dataset[negative_dataset_same_proportion.columns]

In [41]:
# Combine the two datasets
combined_dataset = pd.concat([positive_dataset, negative_dataset_same_proportion])
# shuffle combined_dataset
combined_dataset = combined_dataset.sample(frac=1).reset_index(drop=True)
combined_dataset.head()

CDR3_alfa    TRAV    TRAJ        CDR3_beta      TRBV     TRBJ  \
0             NaN     NaN     NaN   CASSRGAVPNEQFF   TRBV4-1  TRBJ2-1   
1  CAVSGWGSARQLTF  TRAV21  TRAJ22  CASSLPGRFDNEQFF    TRBV27  TRBJ2-1   
2             NaN     NaN     NaN   CASSSRIGTYGYTF   TRBV7-9  TRBJ1-2   
3             NaN     NaN     NaN     CSAFDRVYGYTF  TRBV20-1  TRBJ1-2   
4             NaN     NaN     NaN   CASSPRGPSTEAFF    TRBV19  TRBJ1-1   

   reaction  
0         0  
1         0  
2         1  
3         1  
4         1

In [42]:
def process_gene(g):
    # Code based on code snippet sent by Ceder (on Slack)
    """
    Extracts the gene family, version and allele from a gene in the VDJdb.
    Forward slashes are removed (to make e.g. TRAV14/DV4 and TRAV14DV4 equal), make sure only 1 gene is given as input
    The allele is set to 01 if it is not specified.

    If g is not a string, this functions returns (g, g). This can be used for missing chains when g is None or NaN

    :param g: TRAV, TRAJ, TRBV or TRBJ gene
    :return: the gene formatted as {family}-{version}*{allele} and the family
    """
    # If the gene is None or NaN, return the same as the family
    if not isinstance(g, str):
        return g, g

    # Remove all forward slashes to make e.g. TRAV14/DV4 and TRAV14DV4 equal.
    # g = g.replace('/', '') # TODO: why did I do this? It's causing problems with the tcrdist package, so comment out the line

    # Extract the allele (and set to 01 if none given)
    # Note: Allele already removed, since it wasn't in the background dataset
    # g_allele = g.split('*')
    # if len(g_allele) == 1:
    #     allele = '01'
    # elif len(g_allele) == 2:
    #     g = g_allele[0]
    #     allele = g_allele[1]
    # else:
    #     raise RuntimeError("splitting full gene on '*' resulted in >2 items")

    # Extract the version (and set to 1 if none given)
    group_version = g.split('-')
    if len(group_version) == 1:
        version = '1'
    elif len(group_version) == 2:
        g = group_version[0]
        version = group_version[1]
    else:
        raise RuntimeError("splitting full gene on '-' resulted in >2 items")

    return g, version

# Example:
# TRAV38-2/DV8*01
# Get's converted to:
# TRAV38 2DV8 (and the allele is removed, since it isn't in the negative dataset)

# Process the genes
# for col in ['TRAJ', 'TRAV', 'TRBV', 'TRBJ']:
    # process gene returns the gene and the version, create two new columns for this
    # combined_dataset[col + '_family'], combined_dataset[col + '_version'] = zip(*combined_dataset[col].apply(process_gene))
    # Remove the original column
    # combined_dataset.drop(col, axis=1, inplace=True)

In [43]:
# Create a train and test split
from sklearn.model_selection import train_test_split
train, test = train_test_split(combined_dataset, test_size=0.2, random_state=42)
# Save the combined dataset
test.to_csv('data/generated_combined_dataset_test.csv', index=False)
train.to_csv('data/generated_combined_dataset_train.csv', index=False)

In [44]:
train.head()

CDR3_alfa      TRAV    TRAJ           CDR3_beta      TRBV  \
2310  CILRGENYGGSQGNLIF  TRAV26-2  TRAJ42     CASSLGRVSYTEAFF   TRBV5-1   
5620                NaN       NaN     NaN  CASSYSRGVAGDTDTQYF   TRBV6-2   
9674        CAVSDPGKLTF   TRAV8-6  TRAJ52        CASLRQGQPQHF  TRBV21-1   
6547    CAGAQGSSNTGKLIF    TRAV27  TRAJ37       CASSIRSAYEQYF    TRBV19   
8240   CAVDIHGGGGADGLTF    TRAV39  TRAJ45       CASSVGTGGLQYF    TRBV27   

         TRBJ  reaction  
2310  TRBJ1-1         0  
5620  TRBJ2-3         0  
9674  TRBJ1-5         0  
6547  TRBJ2-7         1  
8240  TRBJ2-7         0

# Summary
I was given two datasets. One contained 5192 positive samples (42% containing the alpha part, 99.96% containing beta) of TCR cells that react with the GILGFVFTL epitope (Influenza A virus).
The other dataset (20788 samples) could be used as a negative dataset (after filtering out the alphas and betas that occured in the positive dataset and making part of the alphas and betas NaN's, to get the same proportions as the positive dataset). Those two sets where then combined into one large dataframe, which was split into a training and a testing file.

In [4]:
# using vdjdb dataset
from util import get_epitope_df
df = get_epitope_df('GILGFVFTL', species='HomoSapiens') # only human (because tcrdistance only works with one species)
df.head()

C:\Users\arnod\PycharmProjects\thesis\util\epitopes.py:6: DtypeWarning: Columns (20,29,30) have mixed types. Specify dtype option on import or set low_memory=False.
  df_vdjdb = pd.read_csv('data/vdjdb-2022-03-30/vdjdb_full.txt', sep='\t')


Number of negative samples changed from 47033 to 44205 (because of overlap with positive samples)
Positive samples: alpha only: 2580, beta only: 3056, both: 2194, none: 0
Negative samples: alpha only: 4499, beta only: 17510, both: 22196, none: 0


CDR3_alfa           TRAV       TRAJ           CDR3_beta  \
14134                NaN            NaN        NaN  CASSYPRDAGTSLDIQYF   
12294   CALILYSGGGADGLTF      TRAV19*01  TRAJ45*01                 NaN   
13527  CALGVDAGGTSYGKLTF      TRAV19*01  TRAJ52*01                 NaN   
14251                NaN            NaN        NaN     CASSIYSSSYNEQFF   
52619    CAASAANFANDKLTF  TRAV29/DV5*01  TRAJ48*01                 NaN   

             TRBV        TRBJ  reaction TRAJ_family TRAJ_version TRAV_family  \
14134  TRBV6-2*01  TRBJ2-4*01         1         NaN          NaN         NaN   
12294         NaN         NaN         1      TRAJ45            1      TRAV19   
13527         NaN         NaN         1      TRAJ52            1      TRAV19   
14251   TRBV19*01  TRBJ2-1*01         1         NaN          NaN         NaN   
52619         NaN         NaN         0      TRAJ48            1  TRAV29/DV5   

      TRAV_version TRBV_family TRBV_version TRBJ_family TRBJ_version  
14134          NaN       TRBV6            2       TRBJ2            4  
12294            1         NaN          NaN         NaN          NaN  
13527            1         NaN          NaN         NaN          NaN  
14251          NaN      TRBV19            1       TRBJ2            1  
52619            1         NaN          NaN         NaN          NaN

In [5]:
df.shape

(15660, 15)

In [6]:
from sklearn.model_selection import train_test_split
train, test = train_test_split(df, test_size=0.2, random_state=42)
# Save the combined dataset
test.to_csv('data/generated_combined_dataset_test_vdjdb.csv', index=False)
train.to_csv('data/generated_combined_dataset_train_vdjdb.csv', index=False)